### ISSUE TITLE
MathF.ScaleB performance in Windows 10 much worse than Ubuntu 20.04

### ISSUE THREAD

**Issue by alexcovington** [dotnet/runtime#40956]
### Description

When running the `System.MathBenchmarks.Single.ScaleB` and `System.MathBenchmarks.Double.ScaleB` benchmarks from [dotnet/performance](https://github.com/dotnet/performance), there seems to be a significant performance difference between Windows 10 and Ubuntu 20.04 where Windows is taking around 10x as long to complete the benchmark.

I'm running the benchmarks using `benchmarks_ci.py` on both machines with the same arguments:

```
py .\scripts\benchmarks_cip.py -c Release -f netcoreapp5.0 --filter 'System.MathBenchmarks.*.ScaleB' --bdn-arguments '--iterationCount 30'
```

### Configuration

See Benchmark.NET results below. I'm happy to lookup any details about either machine.

### Regression?

No.

### Data

**Linux Results**
```
[2020/08/17 15:32:50][INFO] // * Summary *                                                                                                
[2020/08/17 15:32:50][INFO]                                                                                                               
[2020/08/17 15:32:50][INFO] BenchmarkDotNet=v0.12.1.1405-nightly, OS=ubuntu 20.04                           
[2020/08/17 15:32:50][INFO] AMD Ryzen 5 3600, 1 CPU, 12 logical and 6 physical cores                            
[2020/08/17 15:32:50][INFO] .NET Core SDK=5.0.100-rc.1.20413.9                                
[2020/08/17 15:32:50][INFO]   [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.40416, CoreFX 5.0.20.40416), X64 RyuJIT   
[2020/08/17 15:32:50][INFO]   Job-YEQPTW : .NET Core 5.0.0 (CoreCLR 5.0.20.40416, CoreFX 5.0.20.40416), X64 RyuJIT    
[2020/08/17 15:32:50][INFO]  
[2020/08/17 15:32:50][INFO] PowerPlanMode=00000000-0000-0000-0000-000000000000  Runtime=.NET Core 5.0  Arguments=/p:DebugType=portable    
[2020/08/17 15:32:50][INFO] Toolchain=netcoreapp5.0  IterationCount=30  IterationTime=250.0000 ms
[2020/08/17 15:32:50][INFO] MaxIterationCount=20  MinIterationCount=15  WarmupCount=1
[2020/08/17 15:32:50][INFO]
[2020/08/17 15:32:50][INFO] |   Type | Method |     Mean |    Error |   StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
[2020/08/17 15:32:50][INFO] |------- |------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|------:|------:|----------:|
[2020/08/17 15:32:50][INFO] | Double | ScaleB | 29.34 us | 0.216 us | 0.288 us | 29.54 us | 28.95 us | 29.80 us |     - |     - |     - |         - |
[2020/08/17 15:32:50][INFO] | Single | ScaleB | 27.27 us | 0.001 us | 0.002 us | 27.27 us | 27.26 us | 27.27 us |     - |     - |     - |         - |
[2020/08/17 15:32:50][INFO]
```

**Windows Results**
```
[2020/08/17 15:34:23][INFO] // * Summary *
[2020/08/17 15:34:23][INFO]
[2020/08/17 15:34:23][INFO] BenchmarkDotNet=v0.12.1.1405-nightly, OS=Windows 10.0.19041.450 (2004/May2020Update/20H1)
[2020/08/17 15:34:23][INFO] AMD Ryzen 5 3600, 1 CPU, 12 logical and 6 physical cores
[2020/08/17 15:34:23][INFO] .NET Core SDK=5.0.100-rc.1.20413.9
[2020/08/17 15:34:23][INFO]   [Host]     : .NET Core 5.0.0 (CoreCLR 5.0.20.40416, CoreFX 5.0.20.40416), X64 RyuJIT
[2020/08/17 15:34:23][INFO]   Job-KFXTTL : .NET Core 5.0.0 (CoreCLR 5.0.20.40416, CoreFX 5.0.20.40416), X64 RyuJIT
[2020/08/17 15:34:23][INFO]
[2020/08/17 15:34:23][INFO] PowerPlanMode=00000000-0000-0000-0000-000000000000  Runtime=.NET Core 5.0  Arguments=/p:DebugType=portable
[2020/08/17 15:34:23][INFO] Toolchain=netcoreapp5.0  IterationCount=30  IterationTime=250.0000 ms
[2020/08/17 15:34:23][INFO] MaxIterationCount=20  MinIterationCount=15  WarmupCount=1
[2020/08/17 15:34:23][INFO]
[2020/08/17 15:34:23][INFO] |   Type | Method |     Mean |   Error |  StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
[2020/08/17 15:34:23][INFO] |------- |------- |---------:|--------:|--------:|---------:|---------:|---------:|------:|------:|------:|----------:|
[2020/08/17 15:34:23][INFO] | Double | ScaleB | 275.3 us | 0.55 us | 0.81 us | 275.1 us | 274.2 us | 277.4 us |     - |     - |     - |         - |
[2020/08/17 15:34:23][INFO] | Single | ScaleB | 320.4 us | 0.74 us | 1.09 us | 320.5 us | 318.8 us | 323.1 us |     - |     - |     - |         - |
[2020/08/17 15:34:23][INFO]
```

**Comment by alexcovington** [comment]
I'll add what I've dug through so far, but I may need some help making sense of it. The following will all be focusing on the `System.MathBenchmarks.Single.ScaleB` benchmark but the same holds for `System.MathBenchmarks.Double.ScaleB`.

Here is the benchmark in question:

```c#
public partial class Single
    {
        // Tests MathF.ScaleB(float, int) over 5000 iterations for the domain x: -1, +1; y: +0, +5000

        private const float scaleBDeltaX = -0.0004f;
        private const int scaleBDeltaY = 1;
        private const float scaleBExpectedResult = float.NegativeInfinity;

        [Benchmark]
        public void ScaleB() => ScaleBTest();

         public static void ScaleBTest()
        {
            float result = 0.0f, valueX = -1.0f; 
            int valueY = 0;

            for (int iteration = 0; iteration < MathTests.Iterations; iteration++)
            {
                result += MathF.ScaleB(valueX, valueY);
                valueX += scaleBDeltaX; valueY += scaleBDeltaY;
            }

            float diff = MathF.Abs(scaleBExpectedResult - result);

            if (float.IsNaN(result) || (diff > MathTests.SingleEpsilon))
            {
                throw new Exception($"Expected Result {scaleBExpectedResult,10:g9}; Actual Result {result,10:g9}");
            }
        }
    }
```

And here is the assembly emitted by the JIT on both Windows and Linux:

**Windows**
_.NET Core 5.0.0 (CoreCLR 5.0.20.40416, CoreFX 5.0.20.40416), X64 RyuJIT_
```assembly
; System.MathBenchmarks.Single.ScaleB()
       jmp       near ptr System.MathBenchmarks.Single.ScaleBTest()
; Total bytes of code 5
```
```assembly
; System.MathBenchmarks.Single.ScaleBTest()
       push      rdi
       push      rsi
       push      rbx
       sub       rsp,40
       vzeroupper
       vmovaps   [rsp+30],xmm6
       vmovaps   [rsp+20],xmm7
       vxorps    xmm6,xmm6,xmm6
       vmovss    xmm7,dword ptr [7FFAAB7E2308]
       xor       esi,esi
       xor       edi,edi
M01_L00:
       vmovaps   xmm0,xmm7
       mov       edx,esi
       call      System.MathF.ScaleB(Single, Int32)
       vaddss    xmm6,xmm0,xmm6
       vaddss    xmm7,xmm7,dword ptr [7FFAAB7E230C]
       inc       esi
       inc       edi
       cmp       edi,1388
       jl        short M01_L00
       vmovss    xmm0,dword ptr [7FFAAB7E2310]
       vsubss    xmm0,xmm0,xmm6
       vmovss    xmm1,dword ptr [7FFAAB7E2314]
       vandps    xmm0,xmm0,xmm1
       vucomiss  xmm6,xmm6
       jp        short M01_L01
       vucomiss  xmm0,dword ptr [7FFAAB7E2318]
       ja        short M01_L01
       vmovaps   xmm6,[rsp+30]
       vmovaps   xmm7,[rsp+20]
       add       rsp,40
       pop       rbx
       pop       rsi
       pop       rdi
       ret
M01_L01:
       mov       rcx,offset MT_System.Single
       call      CORINFO_HELP_NEWSFAST
       mov       rsi,rax
       vmovss    xmm0,dword ptr [7FFAAB7E231C]
       vmovss    dword ptr [rsi+8],xmm0
       mov       rcx,offset MT_System.Single
       call      CORINFO_HELP_NEWSFAST
       mov       rdi,rax
       vmovss    dword ptr [rdi+8],xmm6
       mov       rcx,offset MT_System.Exception
       call      CORINFO_HELP_NEWSFAST
       mov       rbx,rax
       mov       ecx,53CD
       mov       rdx,7FFAAB6E95C8
       call      CORINFO_HELP_STRCNS
       mov       rcx,rax
       mov       rdx,rsi
       mov       r8,rdi
       call      System.String.Format(System.String, System.Object, System.Object)
       mov       rsi,rax
       mov       rcx,rbx
       call      System.Exception..ctor()
       lea       rcx,[rbx+10]
       mov       rdx,rsi
       call      CORINFO_HELP_ASSIGN_REF
       mov       rcx,rbx
       call      CORINFO_HELP_THROW
       int       3
; Total bytes of code 271
```

**Linux**
_.NET Core 5.0.0 (CoreCLR 5.0.20.40416, CoreFX 5.0.20.40416), X64 RyuJIT_                
```assembly    
; System.MathBenchmarks.Single.ScaleB()     
       jmp       near ptr System.MathBenchmarks.Single.ScaleBTest()   
; Total bytes of code 5    
``` 
```assembly      
; System.MathBenchmarks.Single.ScaleBTest()
       push      rbp
       push      r15 
       push      r14 
       push      rbx 
       push      rax 
       vzeroupper 
       lea       rbp,[rsp+20] 
       vxorps    xmm0,xmm0,xmm0 
       vmovss    dword ptr [rbp-1C],xmm0 
       vmovss    xmm1,dword ptr [7F5DC93AB400]                    
       xor       ebx,ebx       
       xor       r14d,r14d                                                                                                           
M01_L00:                                                                                                                   
       vmovss    dword ptr [rbp-20],xmm1                                                                                                                                
       vmovaps   xmm0,xmm1                                                                                                                     
       mov       edi,ebx                                                       
       call      System.MathF.ScaleB(Single, Int32)     
       vaddss    xmm0,xmm0,dword ptr [rbp-1C]
       vmovss    xmm1,dword ptr [rbp-20]              
       vaddss    xmm1,xmm1,dword ptr [7F5DC93AB404]                                      
       inc       ebx                                                                                                                         
       inc       r14d                                                                                                             
       cmp       r14d,1388                                                                                                    
       vmovss    dword ptr [rbp-1C],xmm0        
       jl        short M01_L00                                                                                                  
       vmovss    xmm1,dword ptr [7F5DC93AB408]
       vmovss    xmm0,dword ptr [rbp-1C]
       vsubss    xmm1,xmm1,xmm0
       vmovss    xmm2,dword ptr [7F5DC93AB40C]
       vandps    xmm1,xmm1,xmm2
       vmovss    dword ptr [rbp-1C],xmm0
       vucomiss  xmm0,xmm0
       jp        short M01_L01
      vucomiss  xmm1,dword ptr [7F5DC93AB410]
       ja        short M01_L01
       lea       rsp,[rbp-18]
       pop       rbx
       pop       r14
       pop       r15
       pop       rbp
       ret
M01_L01:
       mov       rdi,offset MT_System.Single
       call      CORINFO_HELP_NEWSFAST
       mov       rbx,rax
       vmovss    xmm0,dword ptr [7F5DC93AB414]
       vmovss    dword ptr [rbx+8],xmm0
       mov       rdi,offset MT_System.Single
       call      CORINFO_HELP_NEWSFAST
       mov       r14,rax
       vmovss    xmm0,dword ptr [rbp-1C]
       vmovss    dword ptr [r14+8],xmm0
       mov       rdi,offset MT_System.Exception
       call      CORINFO_HELP_NEWSFAST
       mov       r15,rax
       mov       edi,53B5
       mov       rsi,7F5DC8DC6380
       call      CORINFO_HELP_STRCNS
       mov       rdi,rax
       mov       rsi,rbx
       mov       rdx,r14
       call      System.String.Format(System.String, System.Object, System.Object)
       mov       rbx,rax
       mov       rdi,r15
       call      System.Exception..ctor()
       lea       rdi,[r15+10]
       mov       rsi,rbx
       call      CORINFO_HELP_ASSIGN_REF
       mov       rdi,r15
       call      CORINFO_HELP_THROW
       int       3
; Total bytes of code 295
```

Both cases are producing assembly that makes a call to `System.MathF.ScaleB`. When I look at the implementation of `System.MathF.ScaleB` in `\src\coreclr\src\System.Private.CoreLib\src\System\MathF.CoreCLR.cs`, I see that it has the attribute `[MethodImpl(MethodImplOptions.InternalCall)]`. The docs [here](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.methodimploptions?view=netcore-3.1) explain that this marks the method as an internal method with an implementation in the CLR, however this is where things are getting a little fuzzy for me. Where does that "binding" take place?

**Comment by danmosemsft** [comment]
The Linux discrepancies are discussed here https://github.com/dotnet/runtime/issues/7364

The conclusion IIRC was that it was a variation in design choices made in the libc's.

It sounds like potentially the "ideal" (but costly) fix would be to ship the implementations rather than relying on the various libc's and that's was one of the reasons @tannergooding requested that AMD open source https://github.com/amd/win-libm (you probably have all this context though)

**Comment by alexcovington** [comment]
I see, so it is just reliant on whichever libm is available on the system? I can definitely see that it would be costly to ship and maintain a dedicated implementation with the runtime.

Is it possible to change which libm CoreCLR is built with?

**Comment by danmosemsft** [comment]
I don't think CoreCLR is built with libm. I think it uses whichever is the C runtime for the platform - msvc, glibc, or musl. The idea, as I undestand it, would be to compile one of these implementations (libm may be a convenient one) into CoreCLR itself, and use that on all platforms.

It would probably be best to continue this discussion with @tannergooding who I think is back in about a week or so.

**Comment by alexcovington** [comment]
That makes more sense, thanks for clearing this up. I'll wait for @tannergooding to give his input.

**Comment by tannergooding** [comment]
I don't believe there is a trivial way to adjust the libm bound against today (CC. @jkotas?).
* The closest thing we have is likely the PAL layer, which enables some of its own redirections for various fixups required. I don't believe it is applicable to Windows, however.

The Windows x64 libm implementation is currently mostly https://github.com/amd/win-libm (but is internal and existed prior to AMD open sourcing), but there are some other bits of code for functions that aren't provided by `amd/win-libm` (such as `ldexp` and `scalbn`).
Looking at the  implementation of `scalbn`, it is quite a bit more complicated than you might think, especially given it should be roughly equivalent to `ldexp` when `FLT_RADIX` is 2 (as it is for IEEE 754 `float` and `double`). We could likely fix this particular case by switching over to `ldexp` instead, which has a much simpler implementation (assuming there isn't also some corner case I'm forgetting 😄).

It looks like AMD's latest version of LibM (https://developer.amd.com/amd-aocl/amd-math-library-libm/) does provide `ldexp` and `scalbn` implementations, but unlike `win-libm` that is currently closed source and is also Linux only, so it wouldn't work for Windows, OSX, or ARM (nor could it be ported).

> that it would be costly to ship and maintain a dedicated implementation with the runtime.

There are some open issues tracking further investigation in this area. Possible options include:
* Picking an existing native implementation and using that everywhere
  * amd/win-libm is one option
    * This is missing some functions
    * This is currently archived (read-only), so  a fork would need to exist/be maintained
  * ARM-software/optimized-routines is another
    * This relies heavily on the compiler generating optimal code from C
    * It is also unclear if they would take x86/x64 specific optimizations if they existed and were needed
* Porting an implementation to managed code
  * I had prototyped this with `sinf` and the numbers look promising: https://gist.github.com/tannergooding/103be726d48dc3d0b09e890bad0b892f#file-perf-md
  * However, there is increased maintenance cost and porting cost for this approach, although it likewise comes with its own benefits as well

**Comment by jkotas** [comment]
> I don't believe there is a trivial way to adjust the libm bound against today

On Linux, the dynamic library loader supports overriding libc methods via `LD_PRELOAD`. It is often used to override malloc. There is not anything like that on Windows.

> Windows is taking around 10x as long to complete the benchmark

Should this be filled as performance bug against Microsoft Visual C++? We can always work around it locally, but it does not help in the grand scheme.

**Comment by alexcovington** [comment]
> Should this be filled as performance bug against Microsoft Visual C++?

I can also file this as a bug there too, I was just fuzzy on the details of the implementation in CoreCLR, but it does sound like this isn't a CoreCLR issue directly since the implementation of `scalbn` isn't shipped with the runtime and is dependent on the underlying libc/libm.

> I had prototyped this with `sinf` and the numbers look promising

I'm a little surprised a managed implementation is beating out the native implementation, but that's pretty cool to see.

**Comment by tannergooding** [comment]
> I'm a little surprised a managed implementation is beating out the native implementation, but that's pretty cool to see.

Definitely, although there are a few contributing factors here including interaction with the GC, which should go away with https://github.com/dotnet/runtime/issues/13820.

Being able to automatically use the VEX encoding and not have any checks to switch between the regular and FMA path also helps (since the JIT can detect this at compile time and drop the dead path).

**Comment by alexcovington** [comment]
> Being able to automatically use the VEX encoding and not have any checks to switch between the regular and FMA path also helps (since the JIT can detect this at compile time and drop the dead path).

That seems like a pretty good advantage for x86/AMD64 and I assume ARM, but I'm not as familiar with FMA on ARM.

Curiosity got the best of me, so I took your `sinf` implementation and tweaked it to get a `cosf` implementation and compared the numbers:

|      Method |              f |      Mean |     Error |    StdDev |
|------------ |--------------- |----------:|----------:|----------:|
| ManagedSinf |            NaN |  1.164 ns | 0.0100 ns | 0.0078 ns |
|  NativeSinf |            NaN |  4.897 ns | 0.0029 ns | 0.0025 ns |
|  NativeCosf |            NaN |  4.915 ns | 0.0139 ns | 0.0123 ns |
| ManagedCosf |            NaN |  1.146 ns | 0.0010 ns | 0.0009 ns |
| ManagedSinf |      -Infinity |  1.192 ns | 0.0269 ns | 0.0252 ns |
|  NativeSinf |      -Infinity | 67.966 ns | 0.5249 ns | 0.4910 ns |
|  NativeCosf |      -Infinity | 66.942 ns | 0.5447 ns | 0.4548 ns |
| ManagedCosf |      -Infinity |  1.127 ns | 0.0132 ns | 0.0123 ns |
| ManagedSinf | -3.4028235E+38 | 20.563 ns | 0.1059 ns | 0.0939 ns |
|  NativeSinf | -3.4028235E+38 | 13.073 ns | 0.0557 ns | 0.0494 ns |
|  NativeCosf | -3.4028235E+38 | 17.780 ns | 0.0755 ns | 0.0707 ns |
| ManagedCosf | -3.4028235E+38 | 19.843 ns | 0.2114 ns | 0.1651 ns |
| ManagedSinf |           -0.5 |  2.725 ns | 0.0879 ns | 0.0822 ns |
|  NativeSinf |           -0.5 |  3.174 ns | 0.0034 ns | 0.0030 ns |
|  NativeCosf |           -0.5 |  3.431 ns | 0.0033 ns | 0.0028 ns |
| ManagedCosf |           -0.5 |  2.492 ns | 0.0027 ns | 0.0024 ns |
| ManagedSinf |              0 |  2.231 ns | 0.0073 ns | 0.0064 ns |
|  NativeSinf |              0 |  2.852 ns | 0.0038 ns | 0.0030 ns |
|  NativeCosf |              0 |  2.839 ns | 0.0034 ns | 0.0029 ns |
| ManagedCosf |              0 |  3.316 ns | 0.0426 ns | 0.0398 ns |
| ManagedSinf |            0.5 |  2.689 ns | 0.0845 ns | 0.0940 ns |
|  NativeSinf |            0.5 |  3.177 ns | 0.0065 ns | 0.0054 ns |
|  NativeCosf |            0.5 |  3.436 ns | 0.0117 ns | 0.0104 ns |
| ManagedCosf |            0.5 |  3.025 ns | 0.0470 ns | 0.0440 ns |
| ManagedSinf |  3.4028235E+38 | 20.053 ns | 0.1274 ns | 0.1129 ns |
|  NativeSinf |  3.4028235E+38 | 11.859 ns | 0.0100 ns | 0.0084 ns |
|  NativeCosf |  3.4028235E+38 | 17.790 ns | 0.0397 ns | 0.0331 ns |
| ManagedCosf |  3.4028235E+38 | 19.286 ns | 0.0771 ns | 0.0644 ns |
| ManagedSinf |       Infinity |  1.148 ns | 0.0092 ns | 0.0082 ns |
|  NativeSinf |       Infinity | 68.180 ns | 0.5273 ns | 0.4403 ns |
|  NativeCosf |       Infinity | 67.432 ns | 0.5576 ns | 0.5215 ns |
| ManagedCosf |       Infinity |  1.133 ns | 0.0171 ns | 0.0160 ns |

Undeniably better for edge cases like NaN and infinity, looks to be better for smaller numbers and maybe not so much for larger inputs. Someone has probably already implemented/compared a managed `cosf` though, hopefully I'm not stepping on toes.

I think I'll try implementing `scalbn` in managed code just to see if the numbers are still consistent with the managed `sinf` and `cosf` implementations, assuming it hasn't already been done before.

**Comment by tannergooding** [comment]
AFAIK no one on the .NET team has done it yet. I had only done the initial prototype for `sinf` and had started doing the equivalent work for `sin`.

**Comment by alexcovington** [comment]
Here's what I have for a managed implementation, based on glibc:

```c#
// Based on https://code.woboq.org/userspace/glibc/sysdeps/ieee754/flt-32/s_scalbnf.c.html

const float Two25 = 3.355443200e7f;
const float TwoM25 = 2.9802322388e-8f;
const float Huge = 1.0e30f;
const float Small = 1.0e-30f;

public static float ScaleB(float x, int n)
{
    int ix = BitConverter.SingleToInt32Bits(x);
    int k = (ix & 0x7f800000) >> 23; // Extract exponent
    int sign = float.IsNegative(x) ? -1 : 1;

    if (k == 0)
    {
        // 0 or subnormal
        if ((ix & 0x7fffffff) == 0)
            return x; // +-0

        x *= Two25;
        ix = BitConverter.SingleToInt32Bits(x);
        k = ((ix & 0x7f800000) >> 23) - 25;
    }
    if (k == 0xFF)
    {
        // NaN or Infinity
        return x + x;
    }
    if (n < -50000)
        return Small * Small * sign; // Underflow
    if (n > 50000 || k + n > 0xFE)
        return Huge * Huge * sign; // Overflow

    k = k + n;
    if (k > 0)
    {
        return BitConverter.Int32BitsToSingle((int)((ix & 0x807fffff) | (k << 23)));
    }
    if (k <= -25)
    {
        return Small * Small * sign; // Underflow
    }

    k += 25;
    x = BitConverter.Int32BitsToSingle((int)((ix & 0x807fffff) | (k << 23)));
    return x * TwoM25;
}
```

Which I am pretty sure is correct, I'm still doing some more testing to make sure I have all my edge cases covered. However, the numbers look pretty good so far for a managed implementation (f varies, n fixed to 2 for these tests):

|        Method |              f |      Mean |     Error |    StdDev |
|-------------- |--------------- |----------:|----------:|----------:|
|  NativeScaleB |            NaN |  4.302 ns | 0.0036 ns | 0.0030 ns |
| ManagedScaleB |            NaN |  1.705 ns | 0.0250 ns | 0.0222 ns |
|  NativeScaleB |      -Infinity |  4.893 ns | 0.0134 ns | 0.0112 ns |
| ManagedScaleB |      -Infinity |  1.480 ns | 0.0193 ns | 0.0180 ns |
|  NativeScaleB | -3.4028235E+38 | 64.730 ns | 0.2502 ns | 0.2341 ns |
| ManagedScaleB | -3.4028235E+38 |  2.144 ns | 0.0734 ns | 0.0686 ns |
|  NativeScaleB |           -0.5 |  8.405 ns | 0.1925 ns | 0.1977 ns |
| ManagedScaleB |           -0.5 |  2.579 ns | 0.0075 ns | 0.0066 ns |
|  NativeScaleB |              0 |  4.028 ns | 0.0121 ns | 0.0107 ns |
| ManagedScaleB |              0 |  1.656 ns | 0.0248 ns | 0.0220 ns |
|  NativeScaleB |            0.5 |  8.423 ns | 0.1982 ns | 0.1947 ns |
| ManagedScaleB |            0.5 |  3.258 ns | 0.0144 ns | 0.0120 ns |
|  NativeScaleB |  3.4028235E+38 | 65.243 ns | 0.1646 ns | 0.1459 ns |
| ManagedScaleB |  3.4028235E+38 |  2.535 ns | 0.0073 ns | 0.0061 ns |
|  NativeScaleB |       Infinity |  4.889 ns | 0.0098 ns | 0.0087 ns |
| ManagedScaleB |       Infinity |  1.775 ns | 0.0334 ns | 0.0313 ns |

I want to write some other benchmarks to vary `n`, I just ran out of time for the day so I'll probably update with some other numbers.

**Comment by alexcovington** [comment]
Here are the numbers for the managed C# `ScaleB` on the Linux machine:

| Method |     Mean |    Error |   StdDev |   Median |      Min |      Max | Gen 0 | Gen 1 | Gen 2 | Allocated |
|------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| ScaleB | 17.28 us | 0.029 us | 0.022 us | 17.28 us | 17.26 us | 17.35 us |     - |     - |     - |         - |

Which is also faster than the current CoreCLR implementation on Linux. Seems like it may be worth considering moving to a completely managed `System.Math` and `System.MathF` as a managed implementation seems to close the performance gap significantly and improves performance on both platforms.

**Comment by tannergooding** [comment]
This makes sense to me and I think `ScaleB` might be a good place to start given it is much simpler than most of the other math functions. However, doing so needs sign-off from the others on the internal e-mail thread, such as @jkotas. 

As called out on the same thread, it might be worth checking the perf of a native implementation that is annotated with `SuppressGCTransition`. 
If the implementation is purely in managed then we can't use it for features such as constant folding during the JIT compilation and we'd need to duplicate the logic (as we are doing for `Math.Round`: https://github.com/dotnet/runtime/blob/aa5fdab9654d74bc6274c0b5d820272c8e859621/src/libraries/System.Private.CoreLib/src/System/Math.cs#L867-L874)

**Comment by alexcovington** [comment]
> If the implementation is purely in managed then we can't use it for features such as constant folding during the JIT compilation and we'd need to duplicate the logic

I see, definitely want to avoid duplicating logic, but I don't think I fully follow why the logic is duplicated for `Math.Round`. Specifically, why is the JIT unable to do some optimizations with a managed implementation only?

I'll try playing with the `SuppressGCTransition` attribute to see if there is any perf difference on Windows/Linux and post numbers.

**Comment by jkotas** [comment]
`SuppressGCTransition` is not going to fix the poor implementation of ScaleB in Windows CRT. 

I agree that ScaleB is simple. It is really just bit twiddling, no approximations or magic constants. I think it is fine to rewrite ScaleB in C#.

**Comment by jkotas** [comment]
The tests may need beefing up to cover all interesting cases in the implementation.

**Comment by tannergooding** [comment]
> SuppressGCTransition is not going to fix the poor implementation of ScaleB in Windows CRT.

I meant to have the custom implementation in C/C++ with `SupressGCTransition`, rather than C#.
Hopefully this would allow the better perf and allow the JIT to use the same implementation for constant folding and the like if we decided to enable that in the future.

That being said, I'm unsure if the JIT is restricted from using the managed implementation for the same, which would likely be better overall since the managed implementation can use the VEX encoding and other optimizations, while the native cannot easily.

**Comment by alexcovington** [comment]
>  I think it is fine to rewrite ScaleB in C#.
> The tests may need beefing up to cover all interesting cases in the implementation.

I can put together a PR over the next few days to address these items, and the rest of the math functions can be addressed later.

**Comment by alexcovington** [comment]
Running into an issue with the C# implementation which I believe is an issue with staying IEEE compliant during some intermediate computations.

Here's my current implementation based on [musl](https://git.musl-libc.org/cgit/musl/tree/src/math/scalbnf.c):

```c#
public const int SCALEB_C1 = 0x7f000000; // 0x1p127f
public const int SCALEB_C2 = 0x00800000; // 0x1p-126f
public const int SCALEB_C3 = 0x4b800000; // 0x1p24f

public static float ScaleB(float f, int n)
{
    float y = f;
    if (n > 127)
    {
        y *= BitConverter.Int32BitsToSingle(SCALEB_C1);
        n -= 127;
        if (n > 127)
        {
            y *= BitConverter.Int32BitsToSingle(SCALEB_C1);
            n -= 127;
            if (n > 127)
               n = 127;
        }
    }
    else if (n < -126)
    {
        y *= BitConverter.Int32BitsToSingle(SCALEB_C2) - BitConverter.Int32BitsToSingle(SCALEB_C3);
        n += -126 - 24;
        if (n < -126)
        {
            y *= BitConverter.Int32BitsToSingle(SCALEB_C2) - BitConverter.Int32BitsToSingle(SCALEB_C3);
            n += -126 - 24;
            if (n < -126)
               n = -126;
        }
    }
    float u = BitConverter.Int32BitsToSingle(((int)(0x7f + n) << 23));
    return y * u;
}
```

I then started writing tests based on libc-test ([here's a convenient GitHub fork](https://github.com/AssemblyScript/libc-test)) and found a few cases where the C# implementation is not returning the same values. Right now I only have the [special cases](https://github.com/AssemblyScript/libc-test/blob/master/src/math/special/scalbnf.h) written up and it is failing on the values on lines 15, 17, 18, and 19 of the libc-test scalbnf special cases.

Looking at case 15 specifically, the input values are `0x1p127` (0x7F000000) and `-276` (0xFFFFFEEC), which yields `0x1p-149` (0x00000001) in the musl implementation but gives me positive infinity in C#. It looks like any lines that update `y` in the C# implementation may be rounding incorrectly. @tannergooding, I remember reading an explanation from you about this is another GitHub thread, but I can't find it. Something to do with the order of the rounding can affect the final result.

So my questions are:
- I think the issues are to do with rounding to single floating point precision, but I'm wondering if someone can confirm this?
- Rounding semantics in C# vs IEEE?

Thanks in advance!

**Comment by EgorBo** [comment]
@alexcovington from a quick glance your c# impl is not the same as musl, e.g.
`n += -126 - 24;` vs `n += 126 - 24;` (in two places)

**Comment by tannergooding** [comment]
For the code given, it returning `-infinity` looks correct. I'd guess the issue is elsewhere such as in the difference @EgorBo pointed out.

The math...

The input `f` is `2^127` which is exactly representable as `170,141,183,460,469,231,731,687,303,715,884,105,728`. `n` is then `-276`

`SCALEB_C2` is `2^-126` which is exactly representable as `1.1754943508222875079687365372222456778186655567720875215087517062784172594547271728515625E-38`

`SCALEB_C3` is `2^24` which is exactly representable as `16777216`

`SCALEB_C3 - SCALEB_C2` is then `-1.6777215999999999999999999999999999999999999988245056491777124920312634627777543221813344432279124784912482937215827405452728271484375 × 10^7` which is not exactly representable and rounds to `-16777216` (effectively it was `0 - 2^24` (the neighboring values are `-16777218` and `-16777215`, both of which are farther away from the exact result).

The result is then `2^127 * -2^24`, which is `-2^(127 + 24)` or `-2^151`, which is`-2854495385411919762116571938898990272765493248`. `float.MinValue` is `-340282346638528859811704183484516925440` which contains 7 fewer integral digits and so the result is well outside the range and rounds to `-infinity`.

**Comment by tannergooding** [comment]
It looks like another issue is that `y *= BitConverter.Int32BitsToSingle(SCALEB_C2) - BitConverter.Int32BitsToSingle(SCALEB_C3);` should be `y *= BitConverter.Int32BitsToSingle(SCALEB_C2) * BitConverter.Int32BitsToSingle(SCALEB_C3);` (a multiplication, rather than a subtraction).

**Comment by tannergooding** [comment]
It's also worth noting that the JIT won't currently optimize `BitConverter.Int32BitsToSingle(constant)` and so there will be runtime ops for the above op when it could just be a folded constant instead.

Making them constant floats as the following might be better overall:
```csharp
public const float SCALEB_C1 = 1.7014118E+38f;
public const float SCALEB_C2 = 1.1754944E-38f;
public const float SCALEB_C3 = 16777216f;
```

The literal parser used by the C# compiler should be fully IEEE compliant, as is the .NET Core parsing logic in .NET Core 3.0+, so they should compute to the same bitwise results.

**Comment by alexcovington** [comment]
> @alexcovington from a quick glance your c# impl is not the same as musl, e.g.
> n += -126 - 24; vs n += 126 - 24; (in two places)

Oh, you're right! Glad it's just a typo... I'll fix it and run against the test cases again.

> It's also worth noting that the JIT won't currently optimize BitConverter.Int32BitsToSingle(constant) and so there will be runtime ops for the above op when it could just be a folded constant instead.

I figured it wouldn't, I just wanted to get a correct implementation going before trying to get rid of that. I'll reuse the constants @tannergooding provided, which is much cleaner than using `BitConverter` for each one.

Thanks so much!

**Comment by adamsitnik** [comment]
Fixed by #42476. Big thanks for your help @alexcovington !
